### PR TITLE
Add update-cb3

### PR DESCRIPTION
**Pull request by isuruf**
cc @conda-forge/core

### CODE REVIEW

**Comment by scopatz** [comment]
Thanks @isuruf - can you please add docstrings o this PR

File: conda_smithy/update_cb3.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+        sect = self.section[item]
+        start = sect.lc.line
+        for other in self.section:
+            if other.lc.line > start:
+                end = other.lc.line
+                return Section(sect, start, end)
+        return Section(sect, start, self.end)
+
+
+class NullUndefined(jinja2.Undefined):
```
Comment (CJ-Wright) [inline conda_smithy/update_cb3.py]: Is this in the bot too? Should the bot pull this from here?

File: conda_smithy/update_cb3.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+                       'make', 'msinttypes', 'pkg-config', 'automake-wrapper', 'posix'] \
+                    or req.startswith("{{p") or req.startswith("m2-") \
+                    or (req_rendered in ['perl', 'texlive-core', 'curl', 'openssl', 'tar', 'gzip', 'patch']
+                        and section == 'build' and req_rendered not in reqs['run']):
+                messages['Moving {} from host to build'.format(req)] = True
+                build_lines.append(lines[i].rstrip())
+                change_lines[i] = (lines[i], None)
+                continue
+            if req == 'python' and '# [win]' in line:
+                messages['Moving `python # [win]` which was used for vc matrix'.format(req)] = True
```
Comment (isuruf) [inline conda_smithy/update_cb3.py]: These messages describe what they do. If they aren't enough, I'll add some comments above them.

File: conda_smithy/update_cb3.py [hunk 3]
```diff
@@ -364,9 +371,9 @@ def add_compiler(name, p_name):
 if __name__ == '__main__':
     import argparse
     parser = argparse.ArgumentParser()
-    parser.add_argument("recipe")
-    parser.add_argument("output")
-    parser.add_argument("config")
+    parser.add_argument("recipe", help="Path to recipe meta.yaml")
+    parser.add_argument("output", help="Path where updated recipe is stored")
+    parser.add_argument("config", help="Path to conda_build_config.yaml file")
```
Comment (CJ-Wright) [inline conda_smithy/update_cb3.py]: Is this optional? Can it default to the default conda build yaml for conda forge?

Comment (isuruf) [reply conda_smithy/update_cb3.py]: Yes. This main method is for dev purposes only. Running `conda smithy update-cb3` from the root of the feedstock is enough.

**Comment by ocefpaf** [comment]
LGTM. @isuruf is this ready or do you want to add some more commits?

**Comment by isuruf** [comment]
Needs some more work

**Comment by ocefpaf** [comment]
> Needs some more work

No problem. My understanding is that @CJ-Wright and @justcalamari are planning on using this in the bot, so it would be nice to coordinate with them to get this merged before they re-implement the same on their side.

**Comment by isuruf** [comment]
It's ready from my side. I thought tests were failing before, but it looks like they pass.

**Comment by ocefpaf** [comment]
@CJ-Wright anything else before we merge this?

**Comment by CJ-Wright** [comment]
LGTM. I'm excited to start trying this out.

File: conda_smithy/update_cb3.py [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    if isinstance(url, list):
+        for u in url:
+            r = requests.get(u, allow_redirects=True)
+            if r.ok:
+                url = u
+                break
+    else:
+        r = requests.get(url, allow_redirects=True)
+    fname = url.split('/')[-1]
+    tmp_dir = tempfile.mkdtemp('modernize')
```
Comment (isuruf) [inline conda_smithy/update_cb3.py]: We might need to delete this file as storage is limited in heroku and travis

Comment (isuruf) [reply conda_smithy/update_cb3.py]: I mean after we are done with the folder

Comment (CJ-Wright) [reply conda_smithy/update_cb3.py]: can this use the context manager syntax?

File: conda_smithy/update_cb3.py [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+def update_cb3(recipe_path, conda_build_config_path):
+    '''
+    Update the recipe in `recipe_path` to use conda-build=3 features according
+    to conda-build config yaml in `conda_build_config_path`.
+    Returns the updated recipe and a message about the changes made.
+    '''
+    yaml = ruamel.yaml.YAML()
+    yaml.Constructor = MyConstructor
+    yaml.allow_duplicate_keys = True
+    env = jinja2.Environment(undefined=NullUndefined)
```
Comment (shadowwalkersb) [inline conda_smithy/update_cb3.py]: Does `NullUndefined` need to be imported?

**Comment by shadowwalkersb** [comment]
Tried to use `conda smithy update-cb3` on `pygit2-feedstock`, PR https://github.com/conda-forge/pygit2-feedstock/pull/7. It added C compilers to the recipe and I don't think they are needed. Might be something that needs to be addressed here?

**Comment by ocefpaf** [comment]
This is failing om Python 2.7 and I believe we should just drop support for it altogether. @isuruf are you OK with this?

**Comment by isuruf** [comment]
It's a conda-build issue. https://github.com/conda-forge/conda-smithy/pull/800#issuecomment-397116331

**Comment by ocefpaf** [comment]
> It's a conda-build issue. #800 (comment)

Does not change the fact that we are in 2018 and dropping support for Python 2.7 is a valid question. What do you think @isuruf? People are not going to keep fixing things for 2.7 and we do not have the manpower to send PRs upstream.

Where conda-smithy for 2.7 is needed?

**Comment by msarahan** [comment]
Go ahead and thumbs-down Filipe if you must, John, but your opinion would have much more weight if you were more involved in fixing the bugs, as @isuruf has been.

There is a cost to supporting 2.7.  What is the benefit to it?

**Comment by isuruf** [comment]
My concern is that conda-smithy has to be installed in root environment and even though people use python3, they have a miniconda2 installation. (I moved to a miniconda3 installation, so not an issue for me now)

**Comment by ocefpaf** [comment]
> My concern is that conda-smithy has to be installed in root environment and even though people use python3, they have a miniconda2 installation. (I moved to a miniconda3 installation, so not an issue for me now)

Pain of migrating to Python 3 is unavoidable. The question is: what costs more the effort to keep supporting tools for 2.7 or updating a miniconda installation?

Note that virtually all of the scientific stack (numpy, pandas, xarray) is dropping py2.7 this year and the next! People that waits for 2020 will face a harder curve for migrating.

**Comment by jakirkham** [comment]
> There is a cost to supporting 2.7. What is the benefit to it?

IIRC there are still a few places where we (conda-forge) install miniconda2 (when using conda-smithy) to workaround subtle bugs that we ran into with miniconda3. So it's a matter of which bugs we want to fix. Picking miniconda3 does not avoid bugs entirely unfortunately.

**Comment by jakirkham** [comment]
> Go ahead and thumbs-down Filipe if you must, John, but your opinion would have much more weight if you were more involved in fixing the bugs...

This feels unnecessarily hostile to me. I've shared my opinion (and have now explained it above). Please don't single people out like this in the future. Thanks for understanding.

**Comment by ocefpaf** [comment]
> IIRC there are still a few places where we (conda-forge) install miniconda2

Where? We need to identify those and address them. As you said it is a workaround. Time to fix them!

**Comment by scopatz** [comment]
As per the issue at hand,

1. I don't think we should merge PRs that don't pass the test suite.
2. I don't think dropping Python 2 is something that we have the luxury of doing as a packaging system util it is truly end-of-life.
3. If we can't fix the tests, then we should skip them on platforms where they are known failures. This is a perfectly legitimate way to get the test suite to pass.

**Comment by ocefpaf** [comment]
> This feels unnecessarily hostile to me. 

Let's avoid taking making personal here. You John is the first one to single me out in issues and I never took it as harsh. Also, @msarahan is correct. You thumbed down without explaining and without offering help.

We need a clear strategy to move to Python 3 to avoid halting conda-forge.

**Comment by mbargull** [comment]
> My concern is that conda-smithy has to be installed in root environment

As of `conda >=4.4` not necessarily :wink: -- but discussing the implications of running `conda` in non`-base` environments can be postponed until support for `conda <4.4` is dropped from smithy.

> There is a cost to supporting 2.7. What is the benefit to it?

IMHO, the benefits are not likely to justify the costs for Python 2 support in the context of Conda build tools. Personally, the earlier PY2 support can be dropped for those tools, the better.
But it would seem a little random to do that in a PR like this one.

**Comment by jakirkham** [comment]
It's been a while; so, unfortunately I don't recall all the details. We have periodically had to switch between the two due to various reasons. So being able to switch as a quick fix in various tooling has been helpful. [Here is one]( https://github.com/conda-forge/staged-recipes/blob/8e2e10c619664e516c62d2ffab6ca3385d5a06f8/.travis_scripts/create_feedstocks#L20 ) that is using Miniconda2. There may be others (again sorry it's been a while since I've looked into this).

**Comment by ocefpaf** [comment]
> I don't think dropping Python 2 is something that we have the luxury of doing as a packaging system util it is truly end-of-life.

Sorry, I don't understand that. conda-smithy could be writing in `go` for all I care and we can still package py2 and py3.

> Here is one that is using Miniconda2.

This is what we need to address. I'll look if this is still necessary.

> There may be others (again sorry it's been a while since I've looked into this).

I don't think that there are many others. Time to find out!

**Comment by ocefpaf** [comment]
> If we can't fix the tests, then we should skip them on platforms where they are known failures. This is a perfectly legitimate way to get the test suite to pass.

Skipping without knowing how miniconda2 is used in conda-forge seems dangerous to me.

**Comment by mbargull** [comment]
> I don't think dropping Python 2 is something that we have the luxury of doing as a packaging system util it is truly end-of-life.

For the "products", i.e., produced packages, yes, that would definitely be "luxury" since there are still many PY2 software and users. But up to discussion are only the build tools :wink: (Bioconda, e.g., only uses PY3).

**Comment by ocefpaf** [comment]
> Bioconda, e.g., only uses PY3

:+1:

**Comment by mbargull** [comment]
For the issue at hand (that I caused, BTW, sorry!):


@msarahan in https://github.com/conda/conda-build/pull/2970#issuecomment-397680698:
> I'll get a release with this out quickly so we can unblock conda-forge.

:+1:

**Comment by scopatz** [comment]
I mean, do we know if there are any conda-smithy users on Py2k?  I do use conda-smithy on my own machine pretty regularly.

**Comment by ocefpaf** [comment]
> I mean, do we know if there are any conda-smithy users on Py2k?

Hard to know but is that the question? Will `numpy` support py2.7 forever just b/c people are using it? Again, we need a strategy to move forward.

Note that this is blocking conda-forge at the moment!

> I do use conda-smithy on my own machine pretty regularly.

I recommend you to update to Python 3 :smile:

**Comment by scopatz** [comment]
@ocefpaf I guess I am curious as to why there is a rush? I am as much of a fan of Python 3 as anyone, as you know :wink: But, we can decide to not support Python 2 without actively removing it from everywhere right this moment. We could instead have a ramp down period where we slowly allow Python 2 tests to fail (and then we skip them). This would have the effect of unblocking this PR and other parts of the conda-forge ecosystem, without having to worry about user-base impacts or go through the effort to actively purge Python 2. 

I know this is more of a negotiated solution (where everyone leaves equally unhappy). But I think the goal here should be to get things moving again.

**Comment by ocefpaf** [comment]
> There may be others (again sorry it's been a while since I've looked into this).

A search shows that there are not many places but the one you already pointed out.

https://github.com/search?l=Shell&q=org%3Aconda-forge+miniconda2&type=Code

Let's move this discussion to https://github.com/conda-forge/staged-recipes/pull/6102

**Comment by jakirkham** [comment]
Ok, thanks for looking into this, @ocefpaf.

**Comment by isuruf** [comment]
Tests pass now. Ready for review

**Comment by ocefpaf** [comment]
Thanks @isuruf! (And @msarahan and @mbargull for the cb fix.)

**Pull request merged**